train.py

```diff
@@ -2,84 +2,135 @@
+
+        X_noise = torch.randn((X_batch_real.size(0), parameters.gan_noise_size)).to(self.device)
+        G_output = self.generator(X_noise)
+
+        D_fake = self.discriminator(G_output.float())
+        D_real = self.discriminator(X_batch_real.float())
+
+        D_loss = -torch.mean(torch.cat((torch.log(D_real + 1e-8),
+                                        torch.log(1 - D_fake + 1e-8))))
+
+        return D_loss
+
+    def generator_loss(self, noise_batch_shape):
+        X_noise = torch.randn(noise_batch_shape).to(self.device)
+        G_output = self.generator(X_noise)
+        D_output = self.discriminator(G_output.float())
+
+        G_loss = -torch.mean(torch.log(D_output + 1e-8))
+
+        return G_loss
+
+
+class WGPGANTrainer(GANTrainer):
+    def discriminator_loss(self, parameters, X_batch_real):
+        X_batch_real = X_batch_real.to(self.device)
+
+        X_noise = torch.randn((parameters.batch_size, parameters.gan_noise_size)).to(self.device)
+        G_output = self.generator(X_noise).detach()
+
+        D_fake = self.discriminator(G_output.float())
+        D_real = self.discriminator(X_batch_real.float())
 
-            X_noise = torch.randn((batch_size, parameters.gan_noise_size), device=device)
-            X_batch_fake = generator(X_noise).detach()
+        epsilon = torch.rand(X_batch_real.shape[0], 1).expand(X_batch_real.size()).to(self.device)
 
-            predict = discriminator(X_batch_real.float())
-            loss = criterion(predict, y_real)
+        G_interpolation = epsilon * X_batch_real.float() + (1 - epsilon) * G_output.float()
+        G_interpolation = torch.autograd.Variable(G_interpolation, requires_grad=True)
+        D_interpolation = self.discriminator(G_interpolation)
 
-            optimizer_d.zero_grad()
-            loss.backward()
-            optimizer_d.step()
+        weight = torch.ones(D_interpolation.size(), device=self.device)
 
-            metric_accum.disc_accum(loss.item(), torch.round(predict), y_real)
+        gradients = torch.autograd.grad(outputs=D_interpolation,
+                                        inputs=G_interpolation,
+                                        grad_outputs=weight,
+                                        only_inputs=True,
+                                        create_graph=True,
+                                        retain_graph=True)[0]
 
-            predict = discriminator(X_batch_fake)
-            loss = criterion(predict, y_fake)
+        grad_penalty = self.kwargs['lambda_'] * torch.mean((gradients.norm(2, dim=1) - 1) ** 2)
 
-            optimizer_d.zero_grad()
-            loss.backward()
-            optimizer_d.step()
+        D_loss = torch.mean(D_fake) - torch.mean(D_real) + grad_penalty
 
-            metric_accum.disc_accum(loss.item(), torch.round(predict), y_fake)
+        return D_loss
 
-            if scheduler_d is not None:
-                scheduler_d.step()
+    def generator_loss(self, noise_size):
```

Comment:
the same as above, this loss function already implemented in torch, isn't it?

Reply:
No.